io/include/detray/io/common/geometry_writer.hpp

```diff
@@ -0,0 +1,256 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/definitions/indexing.hpp"
+#include "detray/io/common/payloads.hpp"
+#include "detray/materials/material_rod.hpp"
+#include "detray/materials/material_slab.hpp"
+
+// System include(s)
+#include <string>
+#include <vector>
+
+namespace detray {
+
+/// @brief Abstract base class for tracking geometry writers
+template <class detector_t>
+class geometry_writer {
```

Comment:
:thinking: Does it make sense to template the entire class on the detector type? Would it not be better to just template the individual functions? Since many of the member functions don't even make use of the `detector_t` type themselves...

Also, I would do all of this without inheritance... :thinking: Yes, the CSV and JSON writers rely on common code for "serializing" the in-memory payloads. That **is** a good design. But they could just use a set of `detray::io::detail::serialize(...)` function overloads for this. They don't need to use a common base class with a virtual interface. Do they? (Do you really want to use polymorphism in the writing?)

Reply:
Yes, that was the plan :) Let me figure out what the design of the final detector writer will be...

Reply:
I added the full ```detector_writer``` pattern in #438 now. Let's discuss there